zone/parcels.cpp

```diff
@@ -0,0 +1,725 @@
+								std::to_string(item_quantity).c_str(),
+								inst->GetItem()->Name,
+								p->second.from_name.c_str()
+							);
+						}
+					}
+					else {
+						MessageString(
+							Chat::Yellow,
+							PARCEL_INV_FULL,
+							merchant->GetCleanName()
+						);
+						SendParcelRetrieveAck();
+						return;
+					}
+				}
+				else if (free_id != INVALID_INDEX) {
+					inst->SetCharges(item_quantity > 0 ? item_quantity : 1);
+					if (PutItemInInventory(free_id, *inst, true)) {
+						MessageString(
+							Chat::Yellow,
+							PARCEL_DELIVERED,
+							merchant->GetCleanName(),
+							inst->GetItem()->Name,
+							p->second.from_name.c_str()
+						);
+					}
+					else {
+						MessageString(
+							Chat::Yellow,
+							PARCEL_INV_FULL,
+							merchant->GetCleanName()
+						);
+						SendParcelRetrieveAck();
+						return;
+					}
+				}
+				else {
+					MessageString(
+						Chat::Yellow,
+						PARCEL_INV_FULL,
+						merchant->GetCleanName()
+					);
+					SendParcelRetrieveAck();
+					return;
+				}
+			}
+		}
+
+		if (player_event_logs.IsEventEnabled(PlayerEvent::PARCEL_RETRIEVE)) {
+			PlayerEvent::ParcelRetrieve e{};
+			e.from_player_name = p->second.from_name;
+			e.item_id          = p->second.item_id;
+			e.quantity         = p->second.quantity;
+			e.sent_date        = p->second.sent_date;
+
+			RecordPlayerEventLog(PlayerEvent::PARCEL_RETRIEVE, e);
+		}
+
+		DeleteParcel(p->second.id);
+		SendParcelDelete(parcel_in);
+		parcels.erase(p);
+	}
+	SendParcelRetrieveAck();
+	SendParcelIconStatus();
+}
+
+bool Client::DeleteParcel(uint32 parcel_id)
+{
+	auto result = BaseParcelsRepository::DeleteOne(database, parcel_id);
+	if (!result) {
+		LogError("Error deleting parcel id {} from the database.", parcel_id);
+		return false;
+	}
+
+	auto it = std::find_if(parcels.cbegin(), parcels.cend(), [&](const auto &x) { return x.second.id == parcel_id; });
+	SetParcelCount(GetParcelCount() - 1);
+
+	return true;
+}
+
+void Client::LoadParcels()
+{
+	parcels.clear();
+	auto results = ParcelsRepository::GetWhere(database, fmt::format("to_name = '{}'", GetCleanName()));
+
+	for (auto const &p: results) {
+		parcels.emplace(p.slot_id, p);
+	}
+
+	SetParcelCount(parcels.size());
+}
+
+void Client::SendParcelDelete(const ParcelRetrieve_Struct parcel_in)
+{
+	auto outapp = new EQApplicationPacket(OP_ShopDeleteParcel, sizeof(ParcelRetrieve_Struct));
+	auto data   = (ParcelRetrieve_Struct *) outapp->pBuffer;
+
+	data->merchant_entity_id = parcel_in.merchant_entity_id;
+	data->player_entity_id   = parcel_in.player_entity_id;
+	data->parcel_slot_id     = parcel_in.parcel_slot_id;
+	data->parcel_item_id     = parcel_in.parcel_item_id;
+
+	FastQueuePacket(&outapp);
+}
+
+
+int32 Client::FindNextFreeParcelSlot(std::string &character_name)
+{
+	auto results = ParcelsRepository::GetWhere(
+		database, fmt::format("to_name = '{}' ORDER BY slot_id ASC", character_name.c_str()));
+
+	if (results.empty()) {
+		return PARCEL_BEGIN_SLOT;
+	}
+	else {
```

Comment:
This else isn't necessary, we already return early

Reply:
Resolved